src/cached_pool/mod.rs

```diff
@@ -0,0 +1,640 @@
+use std::collections::{HashMap, VecDeque};
+use std::hash::BuildHasherDefault;
+use std::fmt;
+use std::sync::{Arc, mpsc};
+use std::thread;
+use std::time::Duration;
+
+lazy_static! {
+    /// The thread pool global event loop on which all thread pools are registered.
+    static ref REGISTRY: Registry = Registry::new();
+}
+
+/// A thread pool that grows automatically as needed, up to a maximum size.
+/// Threads expire after a configurable amount of time.
+#[derive(Clone, Debug)]
+pub struct CachedPool {
+    id: PoolId,
+    registry: Registry,
+    /// Whether the pool should block when dropped until all tasks complete.
+    pub await_termination: bool,
+}
+
+impl CachedPool {
+    /// Create a new thread pool with the given maximum number of threads
+    /// and maximum idle time before thread expiration.
+    #[inline]
+    pub fn new(config: Config) -> CachedPool {
+        REGISTRY.register(config)
+    }
+
+    /// Submit a new task to a thread.
+    ///
+    /// Fails if all threads are busy with tasks, and the thread pool
+    /// is running its maximum configured number of threads.
+    #[inline]
+    pub fn execute<F>(&self, f: F) -> Result<(), Box<Task + Send>>
+        where F: FnOnce() + Send + 'static
+    {
+        self.registry.execute(self.id, f)
+    }
+
+    /// Get debug information about the thread pool.
+    #[inline]
+    pub fn debug(&self) -> DebugInfo {
+        self.registry.debug(self.id)
+    }
+}
+
+impl Drop for CachedPool {
+    fn drop(&mut self) {
+        if self.await_termination {
+            self.registry.deregister_and_await_termination(self.id);
+        } else {
+            self.registry.deregister(self.id);
+        }
+    }
+}
+
+/// Configuration for `CachedPool`.
+#[derive(Clone, Copy, Debug)]
+pub struct Config {
+    /// The maximum number of threads the thread pool can spawn at the same time.
+    pub max_threads: u32,
+    /// The minimum number of threads the thread pool will contain when idle.
+    pub min_threads: u32,
+    /// The amount of time a thread can be idle before expiring.
+    pub max_idle: Duration,
+}
+
+impl Config {
+    /// Creates a new `Config` with the specified max idle time for threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn max_idle(max_idle: Duration) -> Config {
+        Config { max_idle: max_idle, ..Config::default() }
+    }
+
+    /// Creates a new `Config` with the specified maximum number of threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn max_threads(max_threads: u32) -> Config {
+        Config { max_threads: max_threads, ..Config::default() }
+    }
+
+    /// Creates a new `Config` with the specified minimum number of threads, and all other fields
+    /// set to defaults.
+    #[inline]
+    pub fn min_threads(min_threads: u32) -> Config {
+        Config { min_threads: min_threads, ..Config::default() }
+    }
+
+    #[inline]
+    fn max_idle_ms(&self) -> u64 {
```

Comment:
Is the unwrap necessary here? If we're going to unwrap here, maybe add an assert or something when specifying max idle time to make sure it's safe to call max_idle_ms? Also, this function seems to only be used in one place, maybe we can just remove it.